File: scripts/sync-catalog.ts
/**
 * sync-catalog.ts
 * Imports the Abel_Catalog_CLEAN.xlsx Product Master into the Prisma Product table.
 * Also imports BOM Explorer → BomEntry and Category Mapping → category normalization.
 *
 * Usage:  npx tsx scripts/sync-catalog.ts
 *   or:  npx ts-node --skip-project scripts/sync-catalog.ts
 *
 * Flags:
 *   --dry-run     Print stats without writing to DB
 *   --bom-only    Only import BOM relationships
 *   --skip-bom    Skip BOM import
 */

import { PrismaClient } from '@prisma/client'
import * as XLSX from 'xlsx'
import * as path from 'path'
import * as fs from 'fs'

const prisma = new PrismaClient()

// ── Config ──────────────────────────────────────────────────────────
const CATALOG_PATH = path.resolve(__dirname, '../../Abel_Catalog_CLEAN.xlsx')
const DRY_RUN = process.argv.includes('--dry-run')
const BOM_ONLY = process.argv.includes('--bom-only')
const SKIP_BOM = process.argv.includes('--skip-bom')

// ── Types ───────────────────────────────────────────────────────────
interface CatalogRow {
  SKU: string
  'Product Name': string
  'Clean Category': string
  'Original InFlow Category': string
  'Product Type': string
  'Door Size': string | null
  Handing: string | null
  'Core Type': string | null
  'Panel Style': string | null
  'Fire Rating': string | null
  'Jamb Size': string | null
  Casing: string | null
  'Hardware Finish': string | null
  Material: string | null
  'Unit Cost': number
  'Default List Price': number
  'Markup %': number
  'Margin %': number
  UOM: string | null
  Vendor: string | null
  'Is Active': boolean
  Remarks: string | null
}

interface BomRow {
  'Finished Product Name': string
  'Finished Product SKU': string
  'Clean Category': string
  'Component Name': string
  'Component Type': string
  Quantity: number
  UOM: string
}

interface CategoryRow {
  'Original InFlow Category': string
  'Clean Category': string
  'Product Count': number
  'Avg Cost': number
  'Avg Price': number
  Notes: string | null
}

// ── Helpers ─────────────────────────────────────────────────────────
function clean(val: any): string | null {
  if (val === undefined || val === null || val === '') return null
  return String(val).trim()
}

function cleanNum(val: any): number {
  if (val === undefined || val === null || val === '') return 0
  const n = Number(val)
  return isNaN(n) ? 0 : n
}

function mapSubcategory(cleanCategory: string): string | null {
  // Extract subcategory from clean category name
  const parts = cleanCategory.split(' - ')
  return parts.length > 1 ? parts.slice(1).join(' - ') : null
}

function mapCategory(cleanCategory: string): string {
  // Map to the broader portal category
  const cat = cleanCategory.toLowerCase()
  if (cat.includes('exterior door')) return 'Exterior Doors'
  if (cat.includes('interior door') || cat.includes('hollow core') || cat.includes('solid core'))
    return 'Interior Doors'
  if (cat.includes('bifold')) return 'Interior Doors'
  if (cat.includes('pocket')) return 'Interior Doors'
  if (cat.includes('barn door')) return 'Interior Doors'
  if (cat.includes('fire-rated')) return 'Fire-Rated Doors'
  if (cat.includes('patio') || cat.includes('sliding glass')) return 'Patio Doors'
  if (cat.includes('garage')) return 'Exterior Doors'
  if (cat.includes('door frame')) return 'Door Frames'
  if (cat.includes('door slab')) return 'Door Slabs'
  if (cat.includes('trim') || cat.includes('molding') || cat.includes('moulding')) return 'Trim & Molding'
  if (cat.includes('hardware')) return 'Hardware'
  if (cat.includes('jamb')) return 'Jambs'
  if (cat.includes('stair')) return 'Stair Parts'
  if (cat.includes('closet') || cat.includes('shelf')) return 'Closet & Shelf'
  if (cat.includes('lumber') || cat.includes('sheet good')) return 'Lumber & Sheet Goods'
  if (cat.includes('glass') || cat.includes('lite')) return 'Glass & Inserts'
  if (cat.includes('threshold')) return 'Thresholds'
  if (cat.includes('weather')) return 'Weatherstripping'
  if (cat.includes('attic')) return 'Attic Access'
  if (cat.includes('hvac')) return 'HVAC Doors'
  if (cat.includes('service') || cat.includes('labor')) return 'Services & Labor'
  if (cat.includes('building material')) return 'Building Materials'
  if (cat.includes('window')) return 'Window Components'
  if (cat.includes('dunnage')) return 'Dunnage'
  return cleanCategory // fallback to original
}

function normalizeDoorSize(raw: string | null): string | null {
  if (!raw) return null
  // Already formatted like "20\" x 80\"" → convert to compact "2068"
  const match = raw.match(/(\d+)["″]?\s*x\s*(\d+)["″]?/)
  if (match) {
    return `${match[1]}${match[2]}`
  }
  return raw
}

function normalizeCasing(raw: string | null): string | null {
  if (!raw) return null
  const c = raw.toLowerCase()
  if (c.includes('a-col') || c.includes('a-colonial') || c.includes('2-1/4'))
    return 'A-Col'
  if (c.includes('colonial') || c.includes('3-1/4') || c.includes('c-322'))
    return 'C-322'
  if (c.includes('no casing') || c.includes('none')) return null
  return raw
}

function normalizeHardwareFinish(raw: string | null): string | null {
  if (!raw) return null
  const f = raw.toLowerCase()
  if (f.includes('satin nickel') || f === 'sn') return 'SN'
  if (f.includes('oil rubbed bronze') || f === 'orb') return 'ORB'
  if (f.includes('black') || f === 'blk') return 'BLK'
  if (f.includes('antique brass') || f === 'ab') return 'AB'
  if (f.includes('satin chrome') || f === 'sc') return 'SC'
  if (f.includes('polished chrome') || f === 'pc') return 'PC'
  return raw
}

// ── Product Sync ────────────────────────────────────────────────────
async function syncProducts(rows: CatalogRow[]) {
  console.log(`\n📦 Syncing ${rows.length} products...`)

  let created = 0
  let updated = 0
  let skipped = 0
  let errors = 0

  // Process in batches of 50
  const BATCH_SIZE = 50
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE)

    for (const row of batch) {
      try {
        const sku = clean(row.SKU)
        if (!sku) {
          skipped++
          continue
        }

        const name = clean(row['Product Name']) || sku
        const cleanCategory = clean(row['Clean Category']) || 'Other'
        const cost = cleanNum(row['Unit Cost'])
        const listPrice = cleanNum(row['Default List Price'])

        // Calculate minimum margin from the data
        const margin = listPrice > 0 ? (listPrice - cost) / listPrice : 0.25

        const data = {
          name,
          displayName: name,
          category: mapCategory(cleanCategory),
          subcategory: mapSubcategory(cleanCategory) || clean(row['Product Type']),
          cost: cost,
          basePrice: listPrice > 0 ? listPrice : cost * 1.35, // 35% markup fallback
          minMargin: Math.max(0.10, Math.min(margin, 0.60)), // clamp 10%-60%
          doorSize: normalizeDoorSize(clean(row['Door Size'])),
          handing: clean(row.Handing),
          coreType: clean(row['Core Type']),
          panelStyle: clean(row['Panel Style']),
          jambSize: clean(row['Jamb Size']),
          casingCode: normalizeCasing(clean(row.Casing)),
          hardwareFinish: normalizeHardwareFinish(clean(row['Hardware Finish'])),
          material: clean(row.Material),
          fireRating: clean(row['Fire Rating']),
          active: row['Is Active'] !== false,
          inStock: true, // default; real inventory syncs separately
          inflowCategory: clean(row['Original InFlow Category']),
          lastSyncedAt: new Date(),
        }

        if (DRY_RUN) {
          updated++ // count as would-be upserted
          continue
        }

        await prisma.product.upsert({
          where: { sku },
          create: { sku, ...data },
          update: data,
        })

        // Check if it was create vs update
        updated++
      } catch (err: any) {
        errors++
        if (errors <= 5) {
          console.error(`  ❌ Error on SKU "${row.SKU}": ${err.message}`)
        }
      }
    }

    // Progress bar
    const pct = Math.min(100, Math.round(((i + batch.length) / rows.length) * 100))
    process.stdout.write(`\r  Progress: ${pct}% (${i + batch.length}/${rows.length})`)
  }

  console.log(`\n  ✅ Products synced: ${updated} upserted, ${skipped} skipped, ${errors} errors`)
  return { updated, skipped, errors }
}

// ── BOM Sync ────────────────────────────────────────────────────────
async function syncBom(rows: BomRow[]) {
  console.log(`\n🔧 Syncing ${rows.length} BOM entries...`)

  // First, build a SKU → Product ID lookup from the DB
  const products: Array<{ id: string; sku: string; name: string }> =
    await prisma.$queryRawUnsafe(
      `SELECT id, sku, name FROM "Product" WHERE active = true`
    )

  const skuToId: Record<string, string> = {}
  const nameToId: Record<string, string> = {}
  for (const p of products) {
    skuToId[p.sku] = p.id
    nameToId[p.name.toLowerCase()] = p.id
  }

  let created = 0
  let skipped = 0
  let errors = 0

  // Clear existing BOM entries for a clean sync
  if (!DRY_RUN) {
    const deleted = await prisma.bomEntry.deleteMany({})
    console.log(`  🗑️  Cleared ${deleted.count} existing BOM entries`)
  }

  const BATCH_SIZE = 100
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE)

    for (const row of batch) {
      try {
        const parentSku = clean(row['Finished Product SKU'])
        const componentName = clean(row['Component Name'])
        if (!parentSku || !componentName) {
          skipped++
          continue
        }

        const parentId = skuToId[parentSku]
        if (!parentId) {
          skipped++ // Parent product not in DB
          continue
        }

        // Try to find component by name match
        const componentId = nameToId[componentName.toLowerCase()]
        if (!componentId) {
          skipped++ // Component not found as a product
          continue
        }

        if (DRY_RUN) {
          created++
          continue
        }

        await prisma.bomEntry.create({
          data: {
            parentId,
            componentId,
            quantity: cleanNum(row.Quantity) || 1,
            componentType: clean(row['Component Type']),
          },
        })
        created++
      } catch (err: any) {
        errors++
        if (errors <= 3) {
          console.error(`  ❌ BOM error: ${err.message}`)
        }
      }
    }

    const pct = Math.min(100, Math.round(((i + batch.length) / rows.length) * 100))
    process.stdout.write(`\r  Progress: ${pct}% (${i + batch.length}/${rows.length})`)
  }

  console.log(`\n  ✅ BOM synced: ${created} created, ${skipped} skipped, ${errors} errors`)
  return { created, skipped, errors }
}

// ── Category Validation ─────────────────────────────────────────────
async function validateCategories(mappingRows: CategoryRow[]) {
  console.log(`\n📊 Validating category mapping (${mappingRows.length} entries)...`)

  // Get actual categories from DB
  const dbCategories: Array<{ category: string; count: bigint }> =
    await prisma.$queryRawUnsafe(
      `SELECT category, COUNT(*)::bigint as count FROM "Product" WHERE active = true GROUP BY category ORDER BY count DESC`
    )

  console.log(`\n  Database categories after sync:`)
  for (const cat of dbCategories) {
    console.log(`    ${cat.category}: ${cat.count} products`)
  }

  // Check coverage
  const totalProducts: Array<{ count: bigint }> =
    await prisma.$queryRawUnsafe(`SELECT COUNT(*)::bigint as count FROM "Product"`)
  const activeProducts: Array<{ count: bigint }> =
    await prisma.$queryRawUnsafe(`SELECT COUNT(*)::bigint as count FROM "Product" WHERE active = true`)
  const withCost: Array<{ count: bigint }> =
    await prisma.$queryRawUnsafe(`SELECT COUNT(*)::bigint as count FROM "Product" WHERE cost > 0`)
  const withPrice: Array<{ count: bigint }> =
    await prisma.$queryRawUnsafe(`SELECT COUNT(*)::bigint as count FROM "Product" WHERE "basePrice" > 0`)

  console.log(`\n  📈 Data Quality:`)
  console.log(`    Total products:   ${totalProducts[0].count}`)
  console.log(`    Active products:  ${activeProducts[0].count}`)
  console.log(`    With cost > 0:    ${withCost[0].count}`)
  console.log(`    With price > 0:   ${withPrice[0].count}`)

  // Attribute coverage
  const attrs = ['doorSize', 'handing', 'coreType', 'panelStyle', 'hardwareFinish', 'material', 'jambSize', 'casingCode', 'fireRating']
  console.log(`\n  🏷️  Attribute Coverage:`)
  for (const attr of attrs) {
    const result: Array<{ count: bigint }> =
      await prisma.$queryRawUnsafe(`SELECT COUNT(*)::bigint as count FROM "Product" WHERE "${attr}" IS NOT NULL`)
    console.log(`    ${attr}: ${result[0].count}`)
  }
}

// ── Main ────────────────────────────────────────────────────────────
async function main() {
  console.log('═══════════════════════════════════════════════════')
  console.log('  Abel Lumber Catalog Sync')
  console.log(`  Source: ${CATALOG_PATH}`)
  console.log(`  Mode: ${DRY_RUN ? '🔍 DRY RUN' : '🚀 LIVE'}`)
  console.log('═══════════════════════════════════════════════════')

  // Check file exists
  if (!fs.existsSync(CATALOG_PATH)) {
    console.error(`❌ File not found: ${CATALOG_PATH}`)
    console.error(`   Place Abel_Catalog_CLEAN.xlsx in the project root's parent directory.`)
    process.exit(1)
  }

  // Read XLSX
  const workbook = XLSX.readFile(CATALOG_PATH)
  console.log(`\n📖 Sheets found: ${workbook.SheetNames.join(', ')}`)

  // ── Product Master ──
  if (!BOM_ONLY) {
    const productSheet = workbook.Sheets[workbook.SheetNames[0]] // "Product Master — Clean"
    const productRows: CatalogRow[] = XLSX.utils.sheet_to_json(productSheet)
    console.log(`   Product Master: ${productRows.length} rows`)
    await syncProducts(productRows)
  }

  // ── BOM Explorer ──
  if (!SKIP_BOM) {
    const bomSheetName = workbook.SheetNames.find(s => s.toLowerCase().includes('bom'))
    if (bomSheetName) {
      const bomSheet = workbook.Sheets[bomSheetName]
      const bomRows: BomRow[] = XLSX.utils.sheet_to_json(bomSheet)
      console.log(`   BOM Explorer: ${bomRows.length} rows`)
      await syncBom(bomRows)
    } else {
      console.log('   ⚠️  No BOM sheet found, skipping.')
    }
  }

  // ── Category Mapping (validation only) ──
  const catSheetName = workbook.SheetNames.find(s => s.toLowerCase().includes('category'))
  if (catSheetName) {
    const catSheet = workbook.Sheets[catSheetName]
    const catRows: CategoryRow[] = XLSX.utils.sheet_to_json(catSheet)
    await validateCategories(catRows)
  }

  console.log('\n═══════════════════════════════════════════════════')
  console.log(DRY_RUN ? '  ✅ Dry run complete — no data written' : '  ✅ Catalog sync complete!')
  console.log('═══════════════════════════════════════════════════\n')
}

main()
  .catch((err) => {
    console.error('Fatal error:', err)
    process.exit(1)
  })
  .finally(() => prisma.$disconnect())

File: scripts/sync-catalog.py
#!/usr/bin/env python3
"""
sync-catalog.py
Imports Abel_Catalog_CLEAN.xlsx → Prisma Product table + BOM entries.

Usage:
  python3 scripts/sync-catalog.py                # Full sync
  python3 scripts/sync-catalog.py --dry-run      # Preview only
  python3 scripts/sync-catalog.py --skip-bom     # Skip BOM import
  python3 scripts/sync-catalog.py --bom-only     # Only BOM import
"""

import os, sys, re, json
from datetime import datetime

import openpyxl
import psycopg2
from psycopg2.extras import execute_values
from urllib.parse import urlparse

# ── Config ──────────────────────────────────────────────────────────
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_DIR = os.path.dirname(SCRIPT_DIR)
CATALOG_PATH = os.path.join(os.path.dirname(PROJECT_DIR), "Abel_Catalog_CLEAN.xlsx")

DRY_RUN = "--dry-run" in sys.argv
BOM_ONLY = "--bom-only" in sys.argv
SKIP_BOM = "--skip-bom" in sys.argv


def get_db_url():
    """Read DATABASE_URL from .env file."""
    env_path = os.path.join(PROJECT_DIR, ".env")
    if not os.path.exists(env_path):
        env_path = os.path.join(PROJECT_DIR, ".env.local")
    if not os.path.exists(env_path):
        print("❌ No .env or .env.local found")
        sys.exit(1)

    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if line.startswith("DATABASE_URL="):
                url = line.split("=", 1)[1].strip().strip('"').strip("'")
                return url

    print("❌ DATABASE_URL not found in .env")
    sys.exit(1)


def connect_db():
    url = get_db_url()
    parsed = urlparse(url)
    conn = psycopg2.connect(
        host=parsed.hostname,
        port=parsed.port or 5432,
        dbname=parsed.path.lstrip("/").split("?")[0],
        user=parsed.username,
        password=parsed.password,
        sslmode="require",
    )
    conn.autocommit = False
    return conn


# ── Normalization helpers ───────────────────────────────────────────
def clean(val):
    if val is None or str(val).strip() == "":
        return None
    return str(val).strip()


def clean_num(val):
    if val is None or str(val).strip() == "":
        return 0.0
    try:
        return float(val)
    except (ValueError, TypeError):
        return 0.0


def map_category(clean_cat):
    if not clean_cat:
        return "Other"
    cat = clean_cat.lower()
    if "exterior door" in cat:
        return "Exterior Doors"
    if any(x in cat for x in ["interior door", "hollow core", "solid core", "bifold", "pocket", "barn door"]):
        return "Interior Doors"
    if "fire-rated" in cat:
        return "Fire-Rated Doors"
    if "patio" in cat or "sliding glass" in cat:
        return "Patio Doors"
    if "garage" in cat:
        return "Exterior Doors"
    if "door frame" in cat:
        return "Door Frames"
    if "door slab" in cat:
        return "Door Slabs"
    if "trim" in cat or "molding" in cat or "moulding" in cat:
        return "Trim & Molding"
    if "hardware" in cat:
        return "Hardware"
    if "jamb" in cat:
        return "Jambs"
    if "stair" in cat:
        return "Stair Parts"
    if "closet" in cat or "shelf" in cat:
        return "Closet & Shelf"
    if "lumber" in cat or "sheet good" in cat:
        return "Lumber & Sheet Goods"
    if "glass" in cat or "lite" in cat:
        return "Glass & Inserts"
    if "threshold" in cat:
        return "Thresholds"
    if "weather" in cat:
        return "Weatherstripping"
    if "attic" in cat:
        return "Attic Access"
    if "hvac" in cat:
        return "HVAC Doors"
    if "service" in cat or "labor" in cat:
        return "Services & Labor"
    if "building material" in cat:
        return "Building Materials"
    if "window" in cat:
        return "Window Components"
    if "dunnage" in cat:
        return "Dunnage"
    return clean_cat


def map_subcategory(clean_cat):
    if not clean_cat:
        return None
    parts = clean_cat.split(" - ")
    return parts[1].strip() if len(parts) > 1 else None


def normalize_door_size(raw):
    if not raw:
        return None
    m = re.search(r'(\d+)["\u2033]?\s*x\s*(\d+)', str(raw))
    if m:
        return f"{m.group(1)}{m.group(2)}"
    return str(raw).strip()


def normalize_hardware_finish(raw):
    if not raw:
        return None
    f = str(raw).lower().strip()
    mapping = {
        "satin nickel": "SN", "sn": "SN",
        "oil rubbed bronze": "ORB", "orb": "ORB",
        "black": "BLK", "blk": "BLK", "matte black": "BLK",
        "antique brass": "AB", "ab": "AB",
        "satin chrome": "SC", "sc": "SC",
        "polished chrome": "PC", "pc": "PC",
    }
    return mapping.get(f, str(raw).strip())


def normalize_casing(raw):
    if not raw:
        return None
    c = str(raw).lower().strip()
    if "a-col" in c or "a-colonial" in c or "2-1/4" in c:
        return "A-Col"
    if "colonial" in c or "3-1/4" in c or "c-322" in c:
        return "C-322"
    if "no casing" in c or "none" in c:
        return None
    return str(raw).strip()


def gen_cuid():
    """Generate a cuid-like ID."""
    import random, time
    ts = hex(int(time.time() * 1000))[2:]
    rand = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=12))
    return f"cl{ts}{rand}"


# ── Product Sync ────────────────────────────────────────────────────
def sync_products(cur, rows):
    print(f"\n📦 Syncing {len(rows)} products...")

    upserted = 0
    skipped = 0
    errors = 0

    for i, row in enumerate(rows):
        try:
            sku = clean(row.get("SKU"))
            if not sku:
                skipped += 1
                continue

            name = clean(row.get("Product Name")) or sku
            clean_cat = clean(row.get("Clean Category")) or "Other"
            cost = clean_num(row.get("Unit Cost"))
            list_price = clean_num(row.get("Default List Price"))
            base_price = list_price if list_price > 0 else cost * 1.35
            margin = (list_price - cost) / list_price if list_price > 0 else 0.25
            min_margin = max(0.10, min(margin, 0.60))
            is_active = row.get("Is Active") is not False and row.get("Is Active") is not None

            category = map_category(clean_cat)
            subcategory = map_subcategory(clean_cat) or clean(row.get("Product Type"))
            door_size = normalize_door_size(clean(row.get("Door Size")))
            handing = clean(row.get("Handing"))
            core_type = clean(row.get("Core Type"))
            panel_style = clean(row.get("Panel Style"))
            jamb_size = clean(row.get("Jamb Size"))
            casing_code = normalize_casing(clean(row.get("Casing")))
            hardware_finish = normalize_hardware_finish(clean(row.get("Hardware Finish")))
            material = clean(row.get("Material"))
            fire_rating = clean(row.get("Fire Rating"))
            inflow_cat = clean(row.get("Original InFlow Category"))

            if DRY_RUN:
                upserted += 1
                continue

            # Upsert: INSERT ... ON CONFLICT(sku) UPDATE
            cur.execute(
                """
                INSERT INTO "Product" (
                    id, sku, name, "displayName", category, subcategory,
                    cost, "basePrice", "minMargin",
                    "doorSize", handing, "coreType", "panelStyle",
                    "jambSize", "casingCode", "hardwareFinish", material, "fireRating",
                    active, "inStock", "inflowCategory", "lastSyncedAt",
                    "createdAt", "updatedAt"
                ) VALUES (
                    %s, %s, %s, %s, %s, %s,
                    %s, %s, %s,
                    %s, %s, %s, %s,
                    %s, %s, %s, %s, %s,
                    %s, true, %s, NOW(),
                    NOW(), NOW()
                )
                ON CONFLICT (sku) DO UPDATE SET
                    name = EXCLUDED.name,
                    "displayName" = EXCLUDED."displayName",
                    category = EXCLUDED.category,
                    subcategory = EXCLUDED.subcategory,
                    cost = EXCLUDED.cost,
                    "basePrice" = EXCLUDED."basePrice",
                    "minMargin" = EXCLUDED."minMargin",
                    "doorSize" = EXCLUDED."doorSize",
                    handing = EXCLUDED.handing,
                    "coreType" = EXCLUDED."coreType",
                    "panelStyle" = EXCLUDED."panelStyle",
                    "jambSize" = EXCLUDED."jambSize",
                    "casingCode" = EXCLUDED."casingCode",
                    "hardwareFinish" = EXCLUDED."hardwareFinish",
                    material = EXCLUDED.material,
                    "fireRating" = EXCLUDED."fireRating",
                    active = EXCLUDED.active,
                    "inflowCategory" = EXCLUDED."inflowCategory",
                    "lastSyncedAt" = NOW(),
                    "updatedAt" = NOW()
                """,
                (
                    gen_cuid(), sku, name, name, category, subcategory,
                    cost, base_price, min_margin,
                    door_size, handing, core_type, panel_style,
                    jamb_size, casing_code, hardware_finish, material, fire_rating,
                    is_active, inflow_cat,
                ),
            )
            upserted += 1

        except Exception as e:
            errors += 1
            if errors <= 5:
                print(f"  ❌ Error on SKU '{row.get('SKU')}': {e}")

        # Progress
        if (i + 1) % 200 == 0 or i == len(rows) - 1:
            pct = round((i + 1) / len(rows) * 100)
            print(f"\r  Progress: {pct}% ({i + 1}/{len(rows)})", end="", flush=True)

    print(f"\n  ✅ Products: {upserted} upserted, {skipped} skipped, {errors} errors")
    return upserted, skipped, errors


# ── BOM Sync ────────────────────────────────────────────────────────
def sync_bom(cur, rows):
    print(f"\n🔧 Syncing {len(rows)} BOM entries...")

    # Build SKU → ID and Name → ID lookup
    cur.execute('SELECT id, sku, name FROM "Product" WHERE active = true')
    products = cur.fetchall()
    sku_to_id = {r[1]: r[0] for r in products}
    name_to_id = {r[2].lower(): r[0] for r in products}

    # Clear existing BOMs
    if not DRY_RUN:
        cur.execute('DELETE FROM "BomEntry"')
        print(f"  🗑️  Cleared existing BOM entries")

    created = 0
    skipped = 0
    errors = 0

    for i, row in enumerate(rows):
        try:
            parent_sku = clean(row.get("Finished Product SKU"))
            component_name = clean(row.get("Component Name"))
            if not parent_sku or not component_name:
                skipped += 1
                continue

            parent_id = sku_to_id.get(parent_sku)
            if not parent_id:
                skipped += 1
                continue

            component_id = name_to_id.get(component_name.lower())
            if not component_id:
                skipped += 1
                continue

            if parent_id == component_id:
                skipped += 1  # Can't be own component
                continue

            if DRY_RUN:
                created += 1
                continue

            qty = clean_num(row.get("Quantity")) or 1.0
            uom = clean(row.get("UOM")) or "ea"

            cur.execute(
                """INSERT INTO "BomEntry" (id, "parentProductId", "componentProductId", quantity, unit, "createdAt")
                   VALUES (%s, %s, %s, %s, %s, NOW())""",
                (gen_cuid(), parent_id, component_id, qty, uom),
            )
            created += 1

        except Exception as e:
            errors += 1
            if errors <= 3:
                print(f"  ❌ BOM error: {e}")

        if (i + 1) % 500 == 0 or i == len(rows) - 1:
            pct = round((i + 1) / len(rows) * 100)
            print(f"\r  Progress: {pct}% ({i + 1}/{len(rows)})", end="", flush=True)

    print(f"\n  ✅ BOM: {created} created, {skipped} skipped, {errors} errors")
    return created, skipped, errors


# ── Validation ──────────────────────────────────────────────────────
def validate(cur):
    print("\n📊 Validating data integrity...")

    cur.execute('SELECT category, COUNT(*) as cnt FROM "Product" WHERE active = true GROUP BY category ORDER BY cnt DESC')
    cats = cur.fetchall()
    print("\n  Categories:")
    for cat, cnt in cats:
        print(f"    {cat}: {cnt}")

    cur.execute('SELECT COUNT(*) FROM "Product"')
    total = cur.fetchone()[0]
    cur.execute('SELECT COUNT(*) FROM "Product" WHERE active = true')
    active = cur.fetchone()[0]
    cur.execute('SELECT COUNT(*) FROM "Product" WHERE cost > 0')
    with_cost = cur.fetchone()[0]
    cur.execute('SELECT COUNT(*) FROM "Product" WHERE "basePrice" > 0')
    with_price = cur.fetchone()[0]
    cur.execute('SELECT COUNT(*) FROM "BomEntry"')
    bom_count = cur.fetchone()[0]

    print(f"\n  📈 Summary:")
    print(f"    Total products:  {total}")
    print(f"    Active:          {active}")
    print(f"    With cost > 0:   {with_cost}")
    print(f"    With price > 0:  {with_price}")
    print(f"    BOM entries:     {bom_count}")

    # Attribute coverage
    attrs = ["doorSize", "handing", "coreType", "panelStyle", "hardwareFinish", "material", "jambSize", "casingCode", "fireRating"]
    print(f"\n  🏷️  Attribute Coverage:")
    for attr in attrs:
        cur.execute(f'SELECT COUNT(*) FROM "Product" WHERE "{attr}" IS NOT NULL')
        cnt = cur.fetchone()[0]
        pct = round(cnt / total * 100) if total > 0 else 0
        print(f"    {attr}: {cnt} ({pct}%)")


# ── Main ────────────────────────────────────────────────────────────
def main():
    print("═" * 55)
    print("  Abel Lumber Catalog Sync")
    print(f"  Source: {CATALOG_PATH}")
    print(f"  Mode: {'🔍 DRY RUN' if DRY_RUN else '🚀 LIVE'}")
    print("═" * 55)

    if not os.path.exists(CATALOG_PATH):
        print(f"❌ File not found: {CATALOG_PATH}")
        sys.exit(1)

    # Load XLSX
    wb = openpyxl.load_workbook(CATALOG_PATH, read_only=True, data_only=True)
    print(f"\n📖 Sheets: {wb.sheetnames}")

    def sheet_to_dicts(sheet_name):
        ws = wb[sheet_name]
        rows_iter = ws.iter_rows(values_only=True)
        headers = [str(h) if h else f"col_{i}" for i, h in enumerate(next(rows_iter))]
        result = []
        for row in rows_iter:
            if all(v is None for v in row):
                continue
            result.append(dict(zip(headers, row)))
        return result

    conn = connect_db()
    cur = conn.cursor()

    try:
        # ── Products ──
        if not BOM_ONLY:
            product_rows = sheet_to_dicts(wb.sheetnames[0])
            print(f"   Product Master: {len(product_rows)} rows")
            sync_products(cur, product_rows)
            if not DRY_RUN:
                conn.commit()

        # ── BOM ──
        if not SKIP_BOM:
            bom_sheet = next((s for s in wb.sheetnames if "bom" in s.lower()), None)
            if bom_sheet:
                bom_rows = sheet_to_dicts(bom_sheet)
                print(f"   BOM Explorer: {len(bom_rows)} rows")
                sync_bom(cur, bom_rows)
                if not DRY_RUN:
                    conn.commit()

        # ── Validate ──
        if not DRY_RUN:
            validate(cur)

    except Exception as e:
        conn.rollback()
        print(f"\n❌ Fatal error: {e}")
        raise
    finally:
        cur.close()
        conn.close()
        wb.close()

    print("\n" + "═" * 55)
    print(f"  {'✅ Dry run complete — no data written' if DRY_RUN else '✅ Catalog sync complete!'}")
    print("═" * 55 + "\n")


if __name__ == "__main__":
    main()
